### ISSUE TITLE
Course: Generative AI and LLMs: Architecture and Data Preparation-Module 1

### ISSUE THREAD

**Issue by niyontezejado** [niyontezejado/GenAI#1]
# MODULE 1: GENERATIVE AI ARCHITECTURE

# 0. **Welcome**

Objectives:
- Generative AI models
- LLMs for NLP-bases applications
- Libraries and tools
- Data preparation
- NLP data loaders

Gen AI is changing our lives in multiple ways:

- Completes code automatically
- Crafts music
- Helps in drugs discovery
- Design games

It also helps us to understands natural languages by:

- Generating contextually relevant conversations
- Summarizing documents
- Performing language translation

Opportunities:

- Lucrative careers: generally offer high salaries and financial stability, often in fields like medicine, technology, and business.
- Developing innovative applications
- Employer's to pick

The Figure below clearly shows what will be covered in module 1 and 2.

<p align="center">
  <img src="https://github.com/user-attachments/assets/bba4d10f-1744-4621-9bf3-f1460c1a4a5b" width="45%" />
  <img src="https://github.com/user-attachments/assets/4ca73a6a-e789-493b-9e86-07cf2375bb59" width="45%" />
</p>

# 1. **Significance of Generative AI**

Generative AI: deep learning models that generate text, images and other contents based on the data they are trained on.
They are trained on patterns and structure and be able to produce new and relevant data.

Types of models:

- Text generation: GPT
- Image generation: DALL-E, and GAN
- Audio generation: WaveNet

Applications of generative AI: Present

- Content creation: articles, blogs post, and marketing materials. Visuals and videos.
- Condensing documents
- Language translation
- Chatbots and virtual assistants
- Data analysis

**Notes:** Generative AI have specific applications in these big industries: Healthcare, Finance, Gaming and IT.

# 2. **Generative AI Architecture and Models**

Architectures and models:

- RNNs
- Transformers
- GANs
- VAEs
- Diffusion models

a. RNNs

 RNNs are artificial neural networks that use sequential or time series data. You use them to solve data related problems with a natural order or time based dependencies.

- RNN: Mechanism: 
RNNs differ from regular neural networks due to **_loops_** built into their structure. This loop-based design enables RNNs to remember previous inputs and influence the current input and output. This design is crucial for tasks that deal with sequences such as language modeling. 

![Image](https://github.com/user-attachments/assets/5c8e30b8-5594-4889-878d-f4fc3d036880)

The loop is very important for tasks that deal with sequences such as language and modeling.

- RNN: Fine-tuning

You should fine tune generative AI architectures and models to make them generate accurate and relevant content.

Fine-tuning is adjusting a pretrained model to improve performance for a specific task or dataset.

- RNN: Application:

NLP, Language translation, Speech recognition, Image captioning 

b. Transformers

Transformers are deep learning models that can translate text and speech in near real time. They take data such as words or numbers, and pass it through different layers. Information flows in one direction, starting at the input layer, going through hidden layers, and reaching the output layer. The transformer employs feedback mechanisms to improve accuracy. Transformers are designed with a **self attention mechanism that enables the model to focus on the most important parts of the information** it's looking at, making its understanding and decision-making more efficient. 

- Transformer: Example

GPT: text generation
It produces text that mirrors training data.

c. GAN    (Generative adversarial network)

-  GAN is a generative AI model consisting of two sub models, **a generator and a discriminator.** 
- The generator creates fake samples and sends them to the discriminator. 
- The discriminator checks their authenticity by comparing them with real samples from a domain set. 

![Image](https://github.com/user-attachments/assets/21466a2d-74b0-4d87-9f2a-e0cae5659553)

- GAN: Applications

. Image generation
. Video generation

d. VAE (Variational Autoencoder)

- VAE, operates on an encoder decoder framework. 
- The encoder network first compresses input data into a simplified abstract space that captures essential characteristics. 
- The decoder network then uses this condensed information to recreate the original data. 
- VAEs focus on learning the underlying patterns within the input data, making it possible to create new data samples that share similar characteristics. 
- VAEs represent data using probability distributions in the latent space. 

![Image](https://github.com/user-attachments/assets/87db7465-e5d5-4888-ba04-fa3757dd1845)

- VAE: Mechanism

VAEs focus on learning the underlying patterns within the input data, making it possible to create new data samples that share similar characteristics. VAEs represent data using probability distributions in the latent space. They can produce a range of possible outputs for a given input, reflecting the inherent uncertainty in real world data.

![Image](https://github.com/user-attachments/assets/41646fc4-b964-4a2c-9a58-0e9e99176e77)

e. Diffusion model: a probabilistic generative model

- A diffusion model is trained to generate images by learning how to remove noise or reconstruct examples from its training data that have been distorted beyond recognition. 
- Depending on the prompt, a diffusion model can generate highly creative images based on the statistical properties of its training data.

![Image](https://github.com/user-attachments/assets/f424d738-4f6c-478b-8599-4406e861dbbc)

- Diffusion model: Application

Generating high quality images from noise or low quality input.

### **Quick comparison**

![Image](https://github.com/user-attachments/assets/78ff1a66-e912-45f2-bd87-9c0c209a216a)

### **Relation to reinforcement learning**

- Generative AI models are closely tied to reinforcement learning. 
- Traditional reinforcement learning focuses on how agents like AI systems or robots interact with their environment to maximize rewards. 
- Generative AI models employ reinforcement learning techniques during training to fine tune and optimize their performance for specific tasks.

### **Recap**

![Image](https://github.com/user-attachments/assets/3d35830e-cd91-4f83-af14-2c535e765fe9)

# 3. **Generative AI for NLP**

a. Generative AI for NLP: Evolution

- Rule-based system
- ML based approach
- DL architecture
- Transformer

b. Generative AI for NLP: Advancements

- text summarization
- machine translation
- sentiment analysis
- chatbot conversations

c. LLM

It uses AI and DL with vast data sets

- Training of LLMs: 
        - understanding languages structures and contexts
        - Capturing nuances of human language
        - Facilitating more natural interactions
        - Predicting the next word in a sequence
        - Producing creative content with minimal task-specific training
- Examples of LLM
        - GPT: act as a decoder, adept at generating text, used in chatbots
        - BERT: Utilizes encoder-only transformer architecture, understand the context of a word in a sentence, used in sentiment analysis and question-answering
        - BART and T5: Uses encoder-decoder architecture, versatile for various NLP tasks

NOTE: Most of LLMs are based on the transformers.

- Versatility of LLMs
        - Pre-train for generic purposes
        - Fine-tune for a much smaller data set

- Coution
      - It may generates inaccurate information
      - Needs to address biases

- Some key libraries and tools are: 
      - PyTorch
      - TensorFlow
      - Hugging Face
      - LangChain
      - Pydantic
